drivers/rescontainervbox/main.go

```diff
@@ -0,0 +1,752 @@
+		command.WithTimeout(*t.StopTimeout),
+	)
+	return cmd.Run()
+}
+
+func (t *T) destroy() error {
+	cmd := command.New(
+		command.WithName("VBoxManage"),
+		command.WithVarArgs("controlvm", t.Name, "poweroff"),
+		command.WithLogger(t.Log()),
+		command.WithCommandLogLevel(zerolog.InfoLevel),
+		command.WithStdoutLogLevel(zerolog.InfoLevel),
+		//command.WithStderrLogLevel(zerolog.ErrorLevel),
+		command.WithTimeout(*t.StopTimeout),
+	)
+	_, err := cmd.Output()
+	return err
+}
+
+func (t *T) containerStart(ctx context.Context) error {
+	if !t.hasConfigFile() {
+		return fmt.Errorf("%s not found", t.configFile())
+	}
+	if err := t.start(); err != nil {
+		return err
+	}
+	return nil
+}
+
+func (t T) isVmInVboxCf() (bool, error) {
+	f, err := os.Open("/root/.config/VirtualBox/VirtualBox.xml")
+	defer f.Close()
+	if err != nil {
+		return false, err
+	}
+	doc, err := xmlquery.Parse(f)
+	if err != nil {
+		return false, err
+	}
+	machineEntries, err := xmlquery.QueryAll(doc, "//MachineEntry")
+	if err != nil {
+		return false, err
+	}
+	for _, entry := range machineEntries {
+		src := entry.SelectAttr("src")
+		if src == filepath.Join("/root/VirtualBox VMs", t.Name, t.Name+".vbox") {
+			return true, nil
+		}
+	}
+	return false, nil
+}
+
+func (t T) addVmToVboxCf() error {
+	cmd := command.New(
+		command.WithName("VBoxManage"),
+		command.WithVarArgs("registervm", filepath.Join("/root/VirtualBox VMs", t.Name, t.Name+".vbox")),
+		command.WithLogger(t.Log()),
+		command.WithCommandLogLevel(zerolog.InfoLevel),
+		command.WithStdoutLogLevel(zerolog.InfoLevel),
+		command.WithStderrLogLevel(zerolog.ErrorLevel),
+	)
+
+	return cmd.Run()
+}
+
+func (t *T) Start(ctx context.Context) error {
+	if v, err := t.isVmInVboxCf(); err != nil {
+		return err
+	} else if !v {
+		if err := t.addVmToVboxCf(); err != nil {
+			return err
+		}
+	}
+	if err := t.ApplyPGChain(ctx); err != nil {
+		return err
+	}
+	if v, err := t.isUp(); err != nil {
+		return err
+	} else if v {
+		t.Log().Info().Msgf("container %s is already up", t.Name)
+		return nil
+	}
+	if err := t.containerStart(ctx); err != nil {
+		return err
+	}
+	actionrollback.Register(ctx, func() error {
+		return t.Stop(ctx)
+	})
+	if err := t.waitForUp(); err != nil {
+		return err
+	}
+
+	if _, err := net.LookupIP(t.hostname()); err == nil {
+		if err := t.waitForPing(); err != nil {
+			return err
+		}
+		if err := t.waitForOperational(); err != nil {
+			return err
+		}
+	} else {
+		t.Log().Debug().Msgf("can not do dns resolution for : %s", t.Name)
+	}
+	return nil
+}
+
+func (t T) Stop(ctx context.Context) error {
+	if v, err := t.isDown(); err != nil {
+		return err
+	} else if v {
+		t.Log().Info().Msgf("container %s is already down", t.Name)
+		return nil
+	}
+	if err := t.containerStop(ctx); err != nil {
+		return err
+	}
+	return nil
+}
+
+func (t T) waitForDown() error {
+	t.Log().Info().Dur("timeout", *t.StopTimeout).Msgf("wait for %s shutdown, max duration %s", t.Name, *t.StopTimeout)
```

Comment:
`
	t.Log().Info().Dur("timeout", *t.StopTimeout).Msgf("wait for %s shutdown, max duration %s", t.Name, *t.StopTimeout)
	->
		t.Log().Info().Msgf("wait for %s shutdown, max duration %s", t.Name, *t.StopTimeout)`